File: WordBreakII.cpp
/*
 * Time Limit Exceeded
*/
#include<iostream>
#include<vector>
#include<string>

using namespace std;

class Solution {
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        vector<string> words;
        vector<string> result;
        wordBreak(s, words, wordDict, result);
        
        return result;
    }
    
    void wordBreak(string s, vector<string>& words, vector<string>& wordDict, vector<string>& result){
        if(s==""){
            string sentence="";
            for(int i=0;i<words.size()-1;i++) sentence += words[i]+" ";
            sentence += words.back();
            result.push_back(sentence);
        }
        for(int i=0;i<wordDict.size();i++){
            /*int j;
            for(j=0;j<words.size();j++){
                if(wordDict[i]==words[j]) break;
            }
            if(j<words.size()) continue;*/
            
            if(s.substr(0, wordDict[i].length())==wordDict[i]){
                words.push_back(wordDict[i]);
                wordBreak(s.substr(wordDict[i].length()), words, wordDict, result);
                words.pop_back();
            }
        }
    }
};

int main(){
    string s;
    cin>>s;
    vector<string> wordDict;
    int n;
    cin>>n;
    for(int i=0;i<n;i++){
        string word;
        cin>>word;
        wordDict.push_back(word);
    }
    
    Solution *solution=new Solution();
    vector<string> result = solution->wordBreak(s, wordDict);
    for(int i=0;i<result.size();i++){
        cout<<result[i]<<endl;
    }
    
    return 0;
}

File: WordBreakII.py
class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: List[str]
        """
        memo = {}
        def wordBreak(s):
            result = []
            if len(s) == 0:
                result.append("")
                return result
            if s in memo:
                return memo[s]
            for word in wordDict:
                if s[ : len(word)] == word:
                    #sublist = wordBreak(s[len(word) : ])
                    #result.append((word + ("" if len(sub)==0 else " ") +sub) for sub in wordBreak(s[len(word) : ]))
                    result += [(word + ("" if len(sub)==0 else " ") +sub) for sub in wordBreak(s[len(word) : ])]
            memo[s] = result
            return result
        return wordBreak(s)

if __name__=='__main__':
    s = raw_input()
    n = int(raw_input())
    wordDict = []
    for i in range(n):
        word = raw_input()
        wordDict.append(word)
    solution = Solution()
    print(solution.wordBreak(s, wordDict))
